pyg_lib/ops/__init__.py

```diff
@@ -332,6 +332,84 @@ def index_sort(
     return torch.ops.pyg.index_sort(inputs, max_value)
 
 
+class Softmax(torch.autograd.Function):
+    @staticmethod
+    def forward(
+        ctx,
+        src: Tensor,
+        index: Optional[Tensor] = None,
+        ptr: Optional[Tensor] = None,
+        num_nodes: Optional[int] = None,
+        dim: int = 0,
+    ) -> Tensor:
+        out = torch.ops.pyg.softmax_forward(src, index, ptr, num_nodes, dim)
+        ctx.save_for_backward(out, index, ptr)
+        ctx.num_nodes = num_nodes
+        ctx.dim = dim
+
+        return out
+
+    @staticmethod
+    def backward(ctx, out_grad: Tensor) -> Tuple[Union[Tensor, int]]:
+        out, index, ptr = ctx.saved_tensors
+        in_grad = torch.ops.pyg.softmax_backward(out, out_grad, index, ptr,
+                                                 ctx.num_nodes, ctx.dim)
+
+        return in_grad, None, None, None, None
+
+
+def softmax(
+    src: Tensor,
+    index: Optional[Tensor] = None,
+    ptr: Optional[Tensor] = None,
+    num_nodes: Optional[int] = None,
+    dim: int = 0,
+) -> Tensor:
+    r"""Computes a sparsely evaluated softmax.
+    Given a value tensor :attr:`src`, this function first groups the values
+    along the given dimension :attr:`dim`, based on the indices specified in
+    :attr:`index`, and then proceeds to compute the softmax individually for
+    each group.
+
+    .. note::
+
+        This operation is currently implemented only for 2D data, where
+        segments are created along the first dimension and are defined using
+        ptr.
+
+    Args:
+        src (Tensor): The source tensor.
+        index (LongTensor, optional): The indices of elements for applying the
+            softmax. (default: :obj:`None`)
+        ptr (LongTensor, optional): If given, computes the softmax based on
+            sorted inputs in CSR representation. (default: :obj:`None`)
+        num_nodes (int, optional): The number of nodes, *i.e.*
+            :obj:`max_val + 1` of :attr:`index`. (default: :obj:`None`)
+        dim (int, optional): The dimension in which to normalize.
+            (default: :obj:`0`)
+
+    :rtype: :class:`Tensor`
+
+    Examples:
+
+        >>> src = torch.randn(4, 4)
+        >>> ptr = torch.tensor([0, 4])
+        >>> softmax(src, None, ptr)
+        tensor([[0.0157, 0.0984, 0.1250, 0.4523],
+                [0.1453, 0.2591, 0.5907, 0.2410],
+                [0.0598, 0.2923, 0.1206, 0.0921],
+                [0.7792, 0.3502, 0.1638, 0.2145]])
+    """
+    if src.dim() != 2 or not src.is_cpu or ptr is None or dim != 0:
```

Comment:
I wonder why ptr is optional because if you don't provide it, you get an error.

Reply:
I wanted to make API in its final form, otherwise, each change here would require a change in `pytorch_geometric`. I'll add support for `index` in the near future.

Reply:
@kgajdamo, after rethinking your suggestion, I decided to change the API and create a specialized `softmax_csr` operation that accepts `ptr` only. Rationale:
`torch.compile` gives nice results for `softmax` with groups defined via `index`, hence I don't see a reason to have a specialized kernel for that option.